src/librustdoc/html/render.rs

```diff
@@ -2506,6 +2507,11 @@ fn render_impl(w: &mut fmt::Formatter, cx: &Context, i: &Impl, link: AssocItemLi
                 if !is_static || render_static {
                     let id = derive_id(format!("{}.{}", shortty, name));
                     write!(w, "<h4 id='{}' class='{}'>", id, shortty)?;
+                    let assoc_link = &AssocItemLink::Anchor(Some(id));
+                    let link = match link {
+                        &AssocItemLink::Anchor(None) => assoc_link,
```

Comment:
Could fill in the actual id for the other items/match arms too?

You may want to move the filling-in of the id into a inherent method of `AssocItemLink` then.